[InferAttrs] Mark floating-point libcalls as errno-writing

antoniofrighetto opened the pull request:
Floating-point libcalls are currently conservatively marked as may write any memory. Restrict these to clobber only `errno`.

dtcxzyw commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -554,6 +579,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotThrow(F);
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
IIUC it should be `memory(argmem: read, errnomem: write)`.

antoniofrighetto replied:
Hmm, shouldn’t marking all pointer arguments as readonly be semantically equivalent to having ArgMem: Ref?

dtcxzyw commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -595,6 +621,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     // May throw; "read" is a valid pthread cancellation point.
     Changed |= setRetAndArgsNoUndef(F);
     Changed |= setDoesNotCapture(F, 1);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
It should be `memory(argmem: write, errnomem: write)`.

dtcxzyw reviewed (commented):
Can you double-check libcalls that also access the argument memory?

dtcxzyw commented on llvm/lib/Transforms/IPO/FunctionAttrs.cpp:
```diff
@@ -128,6 +128,15 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
+  // memory, rather than manually inspecting the underlying object.
+  if (isa<CallInst>(UO)) {
+    auto *Callee = cast<CallInst>(UO)->getCalledFunction();
+    if (Callee && Callee->getName() == "__errno_location") {
```
We use `_errno` on Windows: https://godbolt.org/z/T6oanEKKa
See also https://github.com/llvm/llvm-project/blob/fdadef9be33fb3f0844b85c35a63a8a4137382dd/clang/lib/StaticAnalyzer/Checkers/ErrnoModeling.cpp#L43-L50.

antoniofrighetto replied:
Right, clang on Windows links against UCRT, which declares _errno. This has to be taken into account.

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -672,13 +703,15 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotThrow(F);
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
setOnlyAccessesInaccessibleMemOrErrnoMem()?

Also, why is this setOnlyReadsMemory?

antoniofrighetto replied:
Does not modify the content of the path string in input, so it should make sense to have the pointer argument readonly? Agree that all the other functions have OS/file-system side-effects, so they do access inaccessible memory and errno.

antoniofrighetto replied:
Reviewed, thanks.

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -672,13 +703,15 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setRetAndArgsNoUndef(F);
     Changed |= setDoesNotThrow(F);
     Changed |= setDoesNotCapture(F, 0);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
setOnlyAccessesInaccessibleMemOrErrnoMem()?

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -703,13 +737,15 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotCapture(F, 1);
     Changed |= setOnlyReadsMemory(F, 0);
     Changed |= setOnlyReadsMemory(F, 1);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
setOnlyAccessesInaccessibleMemOrErrnoMem()?

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -703,13 +737,15 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setRetDoesNotAlias(F);
     Changed |= setDoesNotCapture(F, 1);
     Changed |= setOnlyReadsMemory(F, 1);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
setOnlyAccessesInaccessibleMemOrErrnoMem()?

efriedma-quic reviewed (commented):
Please take a second look at your markings: in particular, syscalls with side-effects should be treated as modifying inaccessible memory.

antoniofrighetto commented:
Changes overview from The Open Group Base Specifications:
1) `strtol` (and variants), `strxfrm`, `strcoll`, `scanf` (and variants), `printf`/`sprintf` (and variants), `stat` (and variants), `setitimer`, `getitimer`, `read`, `realpath`, `fseek`, `ftell`, `uname`, `perror`: setOnlyAccessesArgMemOrErrnoMem
3) `malloc`, `memalign`, `calloc`, `tmpfile`: setOnlyAccessesInaccessibleMemOrErrnoMem
4) Do access inaccessible memory / may have side-effects / access filesystem / FILE stream memory: `strdup`, `system`, `mkdir`, `mktime`, `realloc`, `open` (and variants), `gettimeofday`, `chmod`, `chown`, `fscanf`, `fprintf`, `fread`, `fgetpos`, `unlink`, `unsetenv`, `utime`, `getpwnam`, `opendir`, `rename`, `rmdir`, `pclose`, `write`: setOnlyAccessesInaccessibleMemOrArgMemOrErrnoMem

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -353,6 +379,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setDoesNotCapture(F, 1);
     Changed |= setOnlyReadsMemory(F, 1);
+    Changed |= setOnlyAccessesArgMemOrErrnoMem(F);
```
strxfrm/strcoll/sprintf/sscanf/etc. access locale; is that inaccessiblemem, or do we need to drop this marking completely?

antoniofrighetto replied:
Right, adding locale-sensitive routines as well. IIUC this should be set as inaccessiblemem: read as they do not change locale.

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -396,6 +431,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotThrow(F);
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesArgMemOrErrnoMem(F, ModRefInfo::Ref);
```
Umm, are you forgetting about stdin?

antoniofrighetto replied:
Right, should be inaccessiblemem: read.

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -596,6 +643,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     // May throw; "read" is a valid pthread cancellation point.
     Changed |= setRetAndArgsNoUndef(F);
     Changed |= setDoesNotCapture(F, 1);
+    Changed |= setOnlyAccessesArgMemOrErrnoMem(F);
```
read also modifies the file descriptor

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -604,11 +652,19 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotThrow(F);
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesArgMemOrErrnoMem(F);
```
realpath accesses the filesystem?

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -617,19 +673,24 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setDoesNotCapture(F, 1);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesArgMemOrErrnoMem(F);
```
readlink accesses the filesystem?

efriedma-quic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -695,6 +759,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotThrow(F);
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesArgMemOrErrnoMem(F, ModRefInfo::Ref);
```
access() accesses the filesystem?

antoniofrighetto replied:
Right, this is being added as inaccessiblemem: read as it only reads the filesystem.

efriedma-quic reviewed (commented):
I stopped halfway; please make a second pass once you fix the noted issues.

antoniofrighetto commented:
Took a whole second iteration and updated the overview above.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -863,18 +934,23 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotCapture(F, 0);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |=
+        setOnlyAccessesInaccessibleOrArgOrErrnoMem(F, MR::ModRef, MR::Ref);
```
Given the existence of `environ`, we certainly can't model the environment as inaccessible memory.

antoniofrighetto replied:
Right, good point.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -555,12 +591,15 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotThrow(F);
     Changed |= setWillReturn(F);
     Changed |= setDoesNotCapture(F, 0);
+    Changed |= setOnlyAccessesInaccessibleOrArgOrErrnoMem(F);
```
mktime also interacts with the tzname, timezone and daylight globals.

nikic reviewed (commented):
I'm pretty apprehensive about the heavy lifting "inaccessible memory" is doing in this patch -- things like locales, files etc are now modelled as inaccessible memory.

 * I'm not very confident there are no non-libcall ways to access all of these even within the defined API.
 * We also have to worry about accesses of implementation details in headers. If any library calls are implemented as inline functions in headers, they may access memory that is logically inaccessible.
 * This further increases the LTO risk, where this inaccessible memory may become accessible. Currently this problem is mostly limited to allocators, but this expands it to all of libc.

TBH I think that all this patch should be doing is change this one line to restrict to errnomem: https://github.com/llvm/llvm-project/blob/661daa22bfb678de3162369a34042a69b548e883/llvm/lib/Transforms/Utils/BuildLibCalls.cpp#L1415

That's the main case we're interested in handling. (Well, that and malloc, but I think that should only be added after the AA support is there.)

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -320,6 +357,7 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setWillReturn(F);
     Changed |= setDoesNotCapture(F, 1);
     Changed |= setOnlyReadsMemory(F, 0);
+    Changed |= setOnlyAccessesArgOrErrnoMem(F);
```
Accesses locale?

nikic replied:
It shouldn't be inaccessible memory -- but if you say this is argmem + errnomem only you're saying that locale doesn't exist at all. We just shouldn't add memory attributes for things that access locale at all.

antoniofrighetto replied:
Oh, that’s very true. Thanks for the clarification!

nikic reviewed (commented):
> Restricted libc functions back to errnomem only; this can be relieved further if needed.

It looks like you still have lots of functions that access locale (like strtol, strxfrm) listed as "argmem or errnomem only"?

nikic reviewed (changes requested):
To start with, please only add errnomem to the FP libcalls, and nothing else.

antoniofrighetto commented:
> To start with, please only add errnomem to the FP libcalls, and nothing else.

Agree, let's begin  with a reduced set of functions first off. Updated, thanks!

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -1294,26 +1278,53 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
+  case LibFunc_isascii:
+  case LibFunc_isdigit:
+  case LibFunc_labs:
+  case LibFunc_llabs:
```
For all othe other ones, these should be setDoesNotAccessMemory, right? As they don't access ernno or anything else.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -1294,26 +1278,53 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setDoesNotCapture(F, 1);
     Changed |= setDoesNotCapture(F, 2);
     Changed |= setWillReturn(F);
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
That's not right, sincos has out parameters. This should be argmem: write + errno.

antoniofrighetto replied:
Admittedly, thought marking out parameters as writeonly sufficed to carry this information. Exposing modref only for ArgMem, as errno is only written here, per below.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -84,6 +85,14 @@ static bool setOnlyAccessesInaccessibleMemory(Function &F) {
+static bool setOnlyAccessesErrnoMemory(Function &F) {
+  if (F.onlyAccessesErrnoMemory())
+    return false;
+  F.setOnlyAccessesErrnoMemory();
```
All of these functions only write errno, not read it. We should preserve that.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -1145,14 +1154,14 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
   case LibFunc_remquo:
   case LibFunc_remquof:
   case LibFunc_remquol:
     Changed |= setDoesNotCapture(F, 2);
```
Need to be careful with this fallthrough as well, remquo has an out param.

antoniofrighetto replied:
Right, moving remquo together with sincos, except the latter has an extra out parameter.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -1206,37 +1209,18 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
-  case LibFunc_ffsll:
   case LibFunc_floor:
   case LibFunc_floorf:
   case LibFunc_floorl:
```
It looks like you dropped some of the calls that don't set errno, but not all of them? E.g. floor and nearbyint are still here.

antoniofrighetto replied:
Ugh, was looking at a [older](https://pubs.opengroup.org/onlinepubs/007904975/functions/floor.html) version of POSIX. Indeed those do not set errno anymore in recent versions. Updated, thanks!

nikic commented on llvm/lib/IR/Function.cpp:
```diff
@@ -922,6 +922,15 @@ void Function::setOnlyAccessesInaccessibleMemory() {
+}
+void Function::setOnlyAccessesErrnoMemory() {
+  setMemoryEffects(getMemoryEffects() &
+                   MemoryEffects::errnoMemOnly(ModRefInfo::Mod));
```
I don't think hardcoding this Mod under an API with this name is okay. Please rebase over https://github.com/llvm/llvm-project/commit/a502c656881c8bc28b49a08c29421059d00aca26 and remove these Function APIs, keeping everything local to BuildLibCalls.

antoniofrighetto replied:
Definitely, thanks for looking into it.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -1145,14 +1166,9 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
   case LibFunc_ldexp:
   case LibFunc_ldexpf:
   case LibFunc_ldexpl:
+    Changed |= setOnlyAccessesErrnoMemory(F);
```
I would also rename this one to setOnlyWritesErrnoMemory(F) for clarity.

nikic commented on llvm/lib/Transforms/Utils/BuildLibCalls.cpp:
```diff
@@ -1149,14 +1169,9 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
   case LibFunc_ldexp:
   case LibFunc_ldexpf:
   case LibFunc_ldexpl:
```
Can merge ldexp with the rest of the FP libcall list? Should also be able to set doesNotThrow and setDoesNotFreeMemory on it...

nikic reviewed (approved):
LGTM

antoniofrighetto committed:
[InferAttrs] Mark floating-point libcalls as `errno`-writing

Floating-point libcalls are currently conservatively marked as
may write any memory. Restrict these to clobber only `errno`.
